Add typed_ast fast parser to mypy

ddfisher opened the pull request:
To use the new fast parser, you install [typed_ast](https://github.com/ddfisher/typed_ast) and make [these changes](https://github.com/ddfisher/typeshed/commit/48b3b260d1daf2c674f325110c30bcdf7ac483f3) to your local typeshed, then run mypy with `--fast-parser`.  Keep in mind that it currently only works on Python 3 code.

This is not yet ready to be merged, as the typeshed being discussed [here](https://github.com/python/typeshed/issues/83) need to be resolved first.

gvanrossum commented on mypy/build.py:
```diff
@@ -120,6 +120,7 @@ def build(sources: List[BuildSource],
           pyversion: Tuple[int, int] = defaults.PYTHON3_VERSION,
           custom_typing_module: str = None,
           implicit_any: bool = False,
+          fast_parser: bool = False,
```
I would like there to be fewer boolean flags, not more. The original pattern is to define a string for each option (e.g. VERBOSE) in build and to add that string to the flags argument when the option is given. Then checking whether a flag is set can be done by asking whether e.g. `VERBOSE in flags`. This way we can support many flags without a proliferation of boolean arguments. You might want to switch to this pattern for implicit_any too.

gvanrossum commented on mypy/build.py:
```diff
@@ -746,7 +751,9 @@ def process(self) -> None:
 
-        if self.id != 'builtins':
+        if self.id == 'builtins':
+            tree.defs.extend(parse.implicit_builtins)
```
There's probably some place where 'None' is created from whole cloth. If you can find that place, it might be easy enough to add True, False, `__debug__` there.

ddfisher commented:
I switched `fast_parser` to a build flag and moved the `True`/`False`/`__debug__` definitions to be by the `None` definition (and now they look considerably better than they did before IMO).

ddfisher commented on mypy/semanal.py:
```diff
@@ -2214,12 +2214,22 @@ def analyze(self, file: MypyFile, fnam: str, mod_id: str) -> None:
+                ('__debug__', bool_type),
+            ]
+
+            for name, typ in literal_types:
```
@JukkaL this works properly with the real builtins but seems to cause some strange errors in unit tests.  Do you know why that might be?

gvanrossum commented:
Can you be more specific about the failures you think are due to that code fragment in semanal.py? The test logs are completely full of tracebacks due to a missing get_line() on None.

FWIW my guess it's due to the tests playing various tricks with custom builtins files. Many test data files contain a directive of the form `[builtins/fixtures/<something>.py]`. I think those fixtures don't define `bool` which is referenced by your code fragment. I presume it's failing in `bool_type = self.sem.named_type('bool')`.

gvanrossum commented:
I find it a little disconcerting that mypy no longer works with the latest version of typeshed (unless one also applies this PR). Could you make a separate PR to address that? It should just add `True, False, __debug__` to the `builtins` namespace.

ddfisher commented:
If you don't think this PR is ready to be merged soon, I'll do that.  My understanding was the main remaining issue was the build failures, which are related to the `True`, `False`, `__debug__` change anyhow.

With regards to those failures:  I'm a bit confused that they're happening, because `True`, etc are only added when the builtins module is imported, which presumably shouldn't be happening in those tests!

ddfisher commented:
Okay, I figured it out.  All the fixtures mypy imports are imported as if they were builtins, so they all need to define bool.  I hope one day we can make mypy fast enough that we'll no longer need these fixtures files -- they've caused me a lot of trouble so far.

ddfisher commented:
The current build failure is caused by some other typeshed change.  Investigating.

gvanrossum commented:
Honestly, I didn't review this PR very thoroughly because of the
failing tests. Given the nature of this PR I expect several more
rounds of review. Please make a separate PR with a small-as-posible
fix for the `True/False/__debug` issue -- it is already causing
problems validating other typeshed PRs (e.g.
https://github.com/python/typeshed/pull/90).

gnprice commented:
Just completed a round of reading through the `typed_ast` code! If that were a PR in itself I'd want to say so on the PR thread since GitHub's UI otherwise can make things an indeterminate stream of individual line-comments, so since it's a repo instead of a PR this thread seems like the best substitute.

(For this PR, though, I don't think any of the details of that code are blockers; so long as this code is properly gated it won't break anything and it's a matter of reading this code itself.)

ddfisher commented:
I think the current blockers for this PR are:
- figuring out what to do about this issue with updating typeshed in #1239
- me rebasing this on #1238 and #1239 (should be pretty fast; will do as soon as #1239 lands)
- Guido giving this a more thorough examination

I think they have to be resolved in that order, too.

ddfisher commented:
Rebased this diff on the latest to mypy.  I think it should be in a mergeable state now!

@gvanrossum can you take look?

gvanrossum commented on mypy/fastparse.py:
```diff
@@ -0,0 +1,744 @@
+
+    def from_operator(self, op):
+        if isinstance(op, typed_ast.Add):
+            return '+'
```
Could this be done faster and more compactly using a dict keyed by th various typed_ast classes?

ddfisher replied:
Hmm.  And doing a lookup on `type(op)`?  Yeah, I think that should work.  I'll try it out.
